src/NSrtm.Core/BiCubicInterpolation/Spline3DCalculator.cs

```diff
@@ -0,0 +1,145 @@
+                    {
+                        values[1][1],
+                        values[1][2],
+                        values[2][1],
+                        values[2][2],
+                        firstDerivativeX[1][0],
+                        firstDerivativeX[1][1],
+                        firstDerivativeX[2][0],
+                        firstDerivativeX[2][1],
+                        firstDerivativeY[1][0],
+                        firstDerivativeY[1][1],
+                        firstDerivativeY[2][0],
+                        firstDerivativeY[2][1],
+                        crossDerivative[0][0],
+                        crossDerivative[0][1],
+                        crossDerivative[1][0],
+                        crossDerivative[1][1]
+                    };
+            return x;
+        }
+
+        private static double getFromPoly(List<double> coefficients, double xPos, double yPos)
+        {
+            double result = 0;
+            for (int i = 0; i < 4; i++)
+            {
+                for (int j = 0; j < 4; j++)
+                {
+                    result += coefficients[i * 4 + j] * Math.Pow(xPos, j) * Math.Pow(yPos, i);
+                }
+            }
+            return result;
+        }
+
+        #region Static Members
+
+        private static readonly int[,] _linearEquationCoefficients =
+
+        {
+            {1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+            {0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+            {-3, 3, 0, 0, -2, -1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+            {2, -2, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+            {0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0},
+            {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0},
+            {0, 0, 0, 0, 0, 0, 0, 0, -3, 3, 0, 0, -2, -1, 0, 0},
+            {0, 0, 0, 0, 0, 0, 0, 0, 2, -2, 0, 0, 1, 1, 0, 0},
+            {-3, 0, 3, 0, 0, 0, 0, 0, -2, 0, -1, 0, 0, 0, 0, 0},
+            {0, 0, 0, 0, -3, 0, 3, 0, 0, 0, 0, 0, -2, 0, -1, 0},
+            {9, -9, -9, 9, 6, 3, -6, -3, 6, -6, 3, -3, 4, 2, 2, 1},
+            {-6, 6, 6, -6, -3, -3, 3, 3, -4, 4, -2, 2, -2, -2, -1, -1},
+            {2, 0, -2, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0},
+            {0, 0, 0, 0, 2, 0, -2, 0, 0, 0, 0, 0, 1, 0, 1, 0},
+            {-6, 6, 6, -6, -4, -2, 4, 2, -3, 3, -3, 3, -2, -1, -2, -1},
+            {4, -4, -4, 4, 2, 2, -2, -2, 2, -2, 2, -2, 1, 1, 1, 1}
+        };
+
+        /// <summary>
+        ///     This subroutine builds bicubic func spline.
+        ///     For more information see https://en.wikipedia.org/wiki/Bicubic_interpolation
+        /// </summary>
+        /// <param name="values">function values,  (4×4)</param>
+        /// <param name="step">Distance between the nodes</param>
+        /// <returns>Func with spline interpolant</returns>
+        public static Func<double, double, double> GetBiCubicSpline([NotNull] IReadOnlyList<IReadOnlyList<double>> values, double step)
```

Comment:
Imho this function should return struct/immutable class with single method for calculating value at given point.

This type would be also used for caching polynomial params.